Fix lease management during pull and export

dmcgowan opened the pull request:
Ensure that leases have a reasonable expiration and are cleaned up during prune

Currently it seems some leases could be leaked as they were created with no expiration and their removal could fail due to using a canceled context.

- Fixes #48909


**- Description for the changelog**

```markdown changelog
containerd image-store: fix partially pulled images not being garbage-collected [moby#48910](https://github.com/moby/moby/pull/48910)
```

dmcgowan committed:
Fix lease management during image operations

Ensure that leases have a reasonable expiration and are
cleaned up during prune

Signed-off-by: Derek McGowan <derek@mcg.dev>

vvoland commented on daemon/containerd/image_prune.go:
```diff
@@ -60,6 +61,22 @@ func (i *ImageService) ImagesPrune(ctx context.Context, fltrs filters.Args) (*im
 		return nil, err
 	}
 
+	// Prune leases
+	leaseManager := i.client.LeasesService()
+	pullLeases, err := leaseManager.List(ctx, `labels."moby/action.pull"`)
+	if err != nil {
+		return nil, err
+	}
+	for i, lease := range pullLeases {
+		opts := []leases.DeleteOpt{}
+		if i == len(pullLeases)-1 {
+			opts = append(opts, leases.SynchronousDelete)
+		}
+		if err := leaseManager.Delete(ctx, lease, opts...); err != nil {
+			return nil, err
+		}
+	}
```
For a follow up - not sure how feasible that is, but it would be nice if we'd be able to determine how much disk space was freed.

vvoland reviewed (approved):
LGTM, thanks!

thaJeztah commented on daemon/containerd/image_pull.go:
```diff
@@ -40,6 +40,27 @@ func (i *ImageService) PullImage(ctx context.Context, baseRef reference.Named, p
+	ctx, done, err := i.client.WithLease(ctx,
+		leases.WithRandomID(),
+		leases.WithLabels(map[string]string{
+			"moby/action.pull":        baseRef.String(),
```
Should we have a `const` for this one? (currently only used in two places, but not sure if we (should) / will be using the same label in other locations? Do we know if BuildKit also sets such labels, and if these are things we should align?

thaJeztah commented on daemon/containerd/image_pull.go:
```diff
@@ -40,6 +40,27 @@ func (i *ImageService) PullImage(ctx context.Context, baseRef reference.Named, p
+		leases.WithRandomID(),
+		leases.WithLabels(map[string]string{
+			"moby/action.pull":        baseRef.String(),
+			"containerd.io/gc.expire": expireAt.Format(time.RFC3339),
```
Hm... I was surprised that these were not exported as either a `const` or as a utility in containerd; should they be? https://github.com/moby/moby/blob/803534fb4e784d3e6bef9846d64ce4960ad1ec80/vendor/github.com/containerd/containerd/metadata/gc.go#L60-L65

I see containerd v2 added some extra documentation to some of these to outline the required format for values set; https://github.com/containerd/containerd/blob/f0ebbd3c8a662656b458e7e1cca059a690d557a7/core/metadata/gc.go#L69-L76

```go
	// labelGCExpire indicates that an object is collectible after the
	// provided time. For image objects, this makes them available to
	// garbage collect when expired, when not provided, image objects
	// are root objects that never expire. For non-root objects such
	// as content or snapshots, these objects will be treated like
	// root objects before their expiration.
	// Expected format is RFC 3339
	labelGCExpire = []byte("containerd.io/gc.expire")
```


So maybe a utility could make most sense;

- `leases.WithGCExpireAt(time.Time)` (or something) to make sure it's in the correct format
- Some of the other ones may require a specific format as well, so _perhaps_ could require a utility as well to validate (instead of a plain `const` for the right label)?

dmcgowan replied:
There is a utility, here it is just to be specific. There is no logic around it though and the format doesn't change, seemed cleaner to just clearly see the labels that are getting set.

thaJeztah replied:
Oh, I see there's indeed a `WithExpiration`; https://github.com/moby/moby/blob/113d3cecea2c0a6181c34489888a8a89de4e1961/vendor/github.com/containerd/containerd/leases/lease.go#L81-L91


Yeah, mostly looking that using a consistent approach may help finding related things. In this case, I just checked where we're setting these, and I found these;

```bash
git grep '\.WithExpiration' -- ':!vendor'
builder/builder-next/adapters/containerimage/pull.go:   ctx, done, err := leaseutil.WithLease(ctx, p.is.LeaseManager, leases.WithExpiration(5*time.Minute), leaseutil.MakeTemporary)
daemon/containerd/image_builder.go:     ctx, _, err := i.client.WithLease(ctx, leases.WithRandomID(), leases.WithExpiration(1*time.Hour))
daemon/containerd/image_builder.go:             leases.WithExpiration(time.Hour*24),
daemon/containerd/image_builder.go:     ctx, release, err := i.client.WithLease(ctx, leases.WithRandomID(), leases.WithExpiration(1*time.Hour))
daemon/containerd/image_commit.go:      ctx, release, err := i.client.WithLease(ctx, leases.WithRandomID(), leases.WithExpiration(1*time.Hour))
daemon/images/image_pull.go:            leases.WithExpiration(24 * time.Hour),
```

Interestingly it looks like for non-containerd image-pulls, we still use 24 hours, but a different label; https://github.com/moby/moby/blob/113d3cecea2c0a6181c34489888a8a89de4e1961/daemon/images/image_pull.go#L118-L125

And for classic builder with containerd as well, and again with a different label; https://github.com/moby/moby/blob/113d3cecea2c0a6181c34489888a8a89de4e1961/daemon/containerd/image_builder.go#L235-L241


Perhaps those should be corrected as well?

dmcgowan replied:
I'm not going to touch the ones outside of daemon/containerd for now, those might be worth looking but not related to the issues we are seeing with the containerd integration. I added a `withLease` method to our own image service that will add the labels and logic that is consistent with pruning.

thaJeztah reviewed (approved):
overall LGTM

left some comments; nothing _really_ blocking (I think), but possibly some for follow-ups. The timeout is the one I'm wondering if we should shorten as part of this PR (but open to input).

dmcgowan commented on daemon/containerd/image_exporter.go:
```diff
@@ -68,19 +68,18 @@ func (i *ImageService) ExportImage(ctx context.Context, names []string, platform
 	}
 	defer func() {
-		if err := leasesManager.Delete(ctx, lease); err != nil {
+		if err := done(context.WithoutCancel(ctx)); err != nil {
```
In the export case there really isn't a need to hold onto anything since even if the operation is canceled, it will need to restart from the beginning on retry. For the pull case, a cancel may want to keep progress so the context error could be respected. We could make this more explicit and check for a context error rather than let the function error from it.

dmcgowan commented on daemon/containerd/image_pull.go:
```diff
@@ -40,6 +40,27 @@ func (i *ImageService) PullImage(ctx context.Context, baseRef reference.Named, p
 	}()
 	out := streamformatter.NewJSONProgressOutput(outStream, false)
 
+	// How long downloaded content will be held after cancellation
+	expireAt := time.Now().Add(8 * time.Hour)
```
It should for partial downloads of the blobs. Any partially unpacked snapshotters should be discarded as I believe we always use a unique identifier for unpack.

I'm adding a const. We can make this configurable as well in the future if we want and make a smaller value. With it static, it really has to cover the upper bound of possible pull times.

dmcgowan committed:
Update containerd image builder to use lease prune label

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Add label on PrepareSnapshot to warn about non-expiring leases

Currently when preparing a snapshot for a container, a lease is used to
hold that snapshot for the lifespan of a container. That is workaround
to preserve the snapshot when a container is recreated, however, the
containerd object should be able to hold this reference itself.

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan commented:
Looks unrelated, can someone kick the re-run?
```
=== Failed
=== FAIL: github.com/docker/docker/integration/networking TestPortMappedHairpinWindows (12.48s)
    nat_windows_test.go:108: assertion failed: error is not nil: Post "http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.48/containers/7a294dab5fdc253f624b2e5202567ade53acdda4394217cb4fd71888e0994404/start": context deadline exceeded
    panic.go:629: assertion failed: error is not nil: Error response from daemon: error while removing network: network clientnet id 1afe2b3927c272b4aea4d92b6b95345f031e76ad126f8651a6af134f10da0483 has active endpoints

DONE 14 tests, 9 skipped, 1 failure in 33.137s
```

cpuguy83 commented on daemon/containerd/leases.go:
```diff
@@ -0,0 +1,51 @@
+
+	ctx = leases.WithLease(ctx, l.ID)
+	return ctx, func() {
+		if ctx.Err() != nil {
```
This seems racey.  Seems like we should always use `context.WithoutCancel` regardless of if the context is cancelled.

dmcgowan replied:
Updated

cpuguy83 commented on daemon/containerd/leases.go:
```diff
@@ -0,0 +1,51 @@
+	pruneLeaseFilter    = `labels."moby/prune.images"`
+)
+
+func (i *ImageService) withLease(ctx context.Context, cancellable bool) (context.Context, func(), error) {
```
Can you help me understand the reason one would set `cancellable` to true?

dmcgowan replied:
Resumable pull

vvoland commented:
```
    build_userns_linux_test.go:132: run produced invalid output: "", expected "/bin/sleep cap_net_bind_service=eip"
--- FAIL: TestBuildUserNamespaceValidateCapabilitiesAreV2 (15.97s)
```

Looks like this test became flaky recently 🤔

thaJeztah commented:
@vvoland yeah, I think it's one of the tests that became flaky related to the runc 1.2 updates, and there were some remaining patches that didn't get merged (not sure if all of them were good though); https://github.com/moby/moby/pull/48160#issuecomment-2237801422

thaJeztah commented:
- opened https://github.com/moby/moby/pull/48940 to apply some of those patches that weren't merged

dmcgowan commented on daemon/containerd/image_pull.go:
```diff
@@ -89,14 +94,7 @@ func (i *ImageService) pullTag(ctx context.Context, ref reference.Named, platfor
 	var outNewImg containerd.Image
 
 	if oldImage.Target.Digest != "" {
-		// Lease the old image content to prevent it from being garbage collected until we keep it as dangling image.
```
I could add a comment to the `withLease` function, but that is the whole point of the lease, it seems redundant to describe what a lease is everywhere it is used.

thaJeztah replied:
Yeah, this one felt like it added additional context to why we're holding on to the _old_ image (which may not be immediately obvious).

thaJeztah reviewed (approved):
LGTM
